containers/src/Data/Sequence/Internal.hs

```diff
@@ -1376,6 +1377,210 @@ applicativeTree n !mSize m = case n of
     emptyTree = pure EmptyT
 
+data RCountMid a = RCountMid
+  !Int  --total size
+  !(Node a)  -- End of the first
+  !Int -- Number of units in the middle
+  !(Node a)  -- Beginning of the last
+
+{-
+We could generalize beforeSeq quite easily to
+
+  beforeSeq :: (a -> c) -> Seq a -> Seq b -> Seq c
+
+This would let us add a rewrite rule
+
+  fmap f xs <* ys  ==>  beforeSeq f xs ys
+
+We don't currently bother because I don't yet know of a practical use for (<*)
+for sequences; a rewrite rule to optimize it seems like extreme overkill.
+-}
+
+beforeSeq :: Seq a -> Seq b -> Seq a
+beforeSeq xs ys = beforeSeq' xs (length ys)
+
+beforeSeq' :: Seq a -> Int -> Seq a
+beforeSeq' xs lenys = case viewl xs of
+  EmptyL -> empty
+  firstx :< xs' -> case viewr xs' of
+    EmptyR -> replicate lenys firstx
+    Seq xs''FT :> lastx -> case lenys of
+      0 -> empty
+      1 -> xs
+      2 ->
+        Seq $ before2FT firstx xs''FT lastx
+      3 ->
+        Seq $ before3FT firstx xs''FT lastx
+      _ -> Seq $ case lenys `quotRem` 3 of
+             (q,0) -> Deep (lenys * length xs) fd3
+               (raptyMiddle 1 fxE lxE lift_elem xs''FT (RCountMid lenys fn3 (q - 2) ln3))
+               ld3
+                   where
+                    lift_elem a = let n3a = n3 a in (n3a, n3a, n3a)
+             (q,1) -> Deep (lenys * length xs) fd2
+               (raptyMiddle 1 fxE lxE lift_elem xs''FT (RCountMid lenys fn2 (q - 1) ln2))
+               ld2
+                   where
+                    lift_elem a = let n2a = n2 a in (n2a, n3 a, n2a)
+             (q,_) -> Deep (lenys * length xs) fd3
+               (raptyMiddle 1 fxE lxE lift_elem xs''FT (RCountMid lenys fn2 (q - 1) ln3))
+               ld2
+                   where
+                    lift_elem a = let n3a = n3 a in (n3a, n3a, n2 a)
+        where
+          fxE = Elem firstx
+          lxE = Elem lastx
+          fd2 = Two fxE fxE
+          fd3 = Three fxE fxE fxE
+          ld2 = Two lxE lxE
+          ld3 = Three lxE lxE lxE
+          fn2 = Node2 2 fxE fxE
+          fn3 = Node3 3 fxE fxE fxE
+          ln2 = Node2 2 lxE lxE
+          ln3 = Node3 3 lxE lxE lxE
+          n3 a = Node3 3 (Elem a) (Elem a) (Elem a)
+          n2 a = Node2 2 (Elem a) (Elem a)
+
+before2FT :: a -> FingerTree (Elem a) -> a -> FingerTree (Elem a)
+before2FT firstx xs lastx =
+                 Deep (size xs * 2 + 4)
+                      (Two ffx ffx)
+                      (mapMulFT 2 (\(Elem x) -> let fx = Elem x in Node2 2 fx fx) xs)
+                      (Two flx flx)
+  where
+    ffx = Elem firstx
+    flx = Elem lastx
+
+before3FT :: a -> FingerTree (Elem a) -> a -> FingerTree (Elem a)
+before3FT firstx xs lastx =
+                 Deep (size xs * 3 + 6)
+                      (Three ffx ffx ffx)
+                      (mapMulFT 3 (\(Elem x) -> let fx = Elem x in Node3 3 fx fx fx) xs)
+                      (Three flx flx flx)
+  where
+    ffx = Elem firstx
+    flx = Elem lastx
+
+raptyMiddle
```

Comment:
Why `rapty`?

Reply:
No particularly good reason. Do you have a better idea? Also, `blippy` and `swizzle` are quite terrible, but I don't know what to call them.

Reply:
Well, `repEachMiddle` maybe, with the `replicateEach` idea in mind.

As for `blippy`, that's a variant of `lift_elem`... I picked `lift`, but more specific names are possible, like `lift_node` or `lift_level` (because it defines the next level of a 2-3 tree from the one below it by combining nodes in 2s and 3s).

Reply:
`blippy` and `swizzle` are also both variants of `replicate`, by an amount that's ``s `mod` sizec`` for the sides and `sizec` in the middle. So maybe `replicateMod`?

Reply:
`swizzle` is basically just the `fill23` for the root of the 2-3 tree. `blippy` is how to lift `fill23` to the next level, yeah.